File: 02_Scripts/Python_Pipeline/run_pipeline.py
# ==============================================================================
# Thalassemia Survey Data Analysis Pipeline
# Python Main Orchestrator Script
# ==============================================================================

import os
import sys
import subprocess

print("======================================================================")
print("Initializing Thalassemia Survey Python Analysis Pipeline...")
print("======================================================================\n")

# Add current folder to path to enable local module imports
current_dir = os.path.dirname(os.path.abspath(__file__))
if current_dir not in sys.path:
    sys.path.append(current_dir)

# --- 1. Package Management & Setup ---
required_packages = {
    "pandas": "pandas",
    "numpy": "numpy",
    "openpyxl": "openpyxl",
    "matplotlib": "matplotlib",
    "seaborn": "seaborn",
    "scipy": "scipy",
    "statsmodels": "statsmodels"
}

print("Checking package dependencies...")
missing_packages = []
for pkg_name, pip_name in required_packages.items():
    try:
        __import__(pkg_name)
    except ImportError:
        missing_packages.append(pip_name)

if missing_packages:
    print(f"The following required packages are missing: {', '.join(missing_packages)}")
    print("Attempting to install missing packages via pip...")
    try:
        subprocess.check_call([sys.executable, "-m", "pip", "install", "--user"] + missing_packages + ["--break-system-packages"])
        print("Installation completed successfully!\n")
    except Exception as e:
        print(f"Error installing packages: {e}")
        print("Please manually install the missing packages using: pip install " + " ".join(missing_packages))
        sys.exit(1)
else:
    print("All required packages are already installed.\n")

# --- 2. Environment Configurations ---
xlsx_path = "/home/dilshan/Desktop/Thallasemia research/Thalassemia_Research.xlsx"

# Relative fallback
if not os.path.exists(xlsx_path):
    xlsx_path = os.path.join(current_dir, "..", "Thalassemia_Research.xlsx")

if not os.path.exists(xlsx_path):
    print(f"CRITICAL ERROR: Could not locate 'Thalassemia_Research.xlsx'. Checked absolute and parent directories.")
    sys.exit(1)

print(f"Dataset located at: {os.path.abspath(xlsx_path)}")

# Output directories setup
output_base = os.path.join(current_dir, "outputs")
csv_dir = os.path.join(output_base, "csv")
plot_dir = os.path.join(output_base, "plots")

os.makedirs(csv_dir, exist_ok=True)
os.makedirs(plot_dir, exist_ok=True)

print("Output directories initialized under './outputs/'.\n")

# --- 3. Run Pipeline Stages ---

# Stage 1: Load and Clean Data
print("----------------------------------------------------------------------")
print("Stage 1: Executing Data Loader & Cleaning...")
try:
    import loader_01
    df = loader_01.load_and_clean_data(xlsx_path)
    print("Stage 1 completed successfully. Cleaned dataframe loaded.\n")
except Exception as e:
    print(f"CRITICAL ERROR during Stage 1: {e}")
    sys.exit(1)

# Stage 2: Knowledge Scoring & Z-Scores
print("----------------------------------------------------------------------")
print("Stage 2: Executing Scoring System...")
try:
    import scoring_02
    df = scoring_02.calculate_scores(df)
    print("Stage 2 completed successfully. Scores integrated into dataframe.\n")
except Exception as e:
    print(f"CRITICAL ERROR during Stage 2: {e}")
    sys.exit(1)

# Stage 3: Descriptive Statistics Summary
print("----------------------------------------------------------------------")
print("Stage 3: Running Descriptive Statistics summaries...")
try:
    import descriptive_03
    descriptive_03.run_descriptive_stats(df, csv_dir)
    print(f"Stage 3 completed successfully. CSV files saved in {csv_dir}.\n")
except Exception as e:
    print(f"CRITICAL ERROR during Stage 3: {e}")
    sys.exit(1)

# Stage 4: Visualizations
print("----------------------------------------------------------------------")
print("Stage 4: Generating plots...")
try:
    import visualizations_04
    visualizations_04.generate_plots(df, plot_dir)
    print(f"Stage 4 completed successfully. Visual plots saved in {plot_dir}.\n")
except Exception as e:
    print(f"CRITICAL ERROR during Stage 4: {e}")
    sys.exit(1)

# Stage 5: Inferential Statistical Analysis
print("----------------------------------------------------------------------")
print("Stage 5: Executing Inferential Statistical Analysis...")
try:
    import inferential_05
    inferential_05.run_inferential_stats(df, csv_dir)
    print(f"Stage 5 completed successfully. Inferential results saved in {csv_dir}.\n")
except Exception as e:
    print(f"CRITICAL ERROR during Stage 5: {e}")
    sys.exit(1)

print("======================================================================")
print("Python Pipeline Executed Successfully!")
print(f"CSV files folder:  {os.path.abspath(csv_dir)}")
print(f"Plot files folder: {os.path.abspath(plot_dir)}")
print("======================================================================\n")

File: 02_Scripts/R_Pipeline/run_pipeline.R
# ==============================================================================
# Thalassemia Survey Data Analysis Pipeline
# Main Orchestrator Script
# ==============================================================================

cat("======================================================================\n")
cat("Initializing Thalassemia Survey Analysis Pipeline...\n")
cat("======================================================================\n\n")

# --- 1. Package Management & Setup ---
required_packages <- c("readxl", "dplyr", "tidyr", "stringr", "ggplot2")

cat("Checking package dependencies...\n")
missing_packages <- required_packages[!(required_packages %in% installed.packages()[, "Package"])]

if (length(missing_packages) > 0) {
  cat("The following packages are missing and will be installed:\n")
  cat(paste("  -", missing_packages, collapse = "\n"), "\n\n")
  # Use a reliable CRAN mirror
  install.packages(missing_packages, repos = "https://cloud.r-project.org")
} else {
  cat("All required packages are already installed.\n\n")
}

# Load packages
invisible(lapply(required_packages, library, character.only = TRUE))

# --- 2. Environment Configurations ---
# Resolve absolute path to data workbook
xlsx_path <- "/home/dilshan/Desktop/Thallasemia research/01_Data/Raw_Data/Thalassemia_Research.xlsx"

if (!file.exists(xlsx_path)) {
  # Fallback to looking in the parent folder if scripts are running from a relative folder
  xlsx_path <- "../../01_Data/Raw_Data/Thalassemia_Research.xlsx"
}

if (!file.exists(xlsx_path)) {
  stop("CRITICAL ERROR: Could not find 'Thalassemia_Research.xlsx' in absolute path or parent folder.\n")
}

cat("Dataset located at:", normalizePath(xlsx_path), "\n")

# Create output directories if they do not exist
output_base <- "./outputs"
csv_dir <- file.path(output_base, "csv")
plot_dir <- file.path(output_base, "plots")

if (!dir.exists(output_base)) dir.create(output_base)
if (!dir.exists(csv_dir)) dir.create(csv_dir)
if (!dir.exists(plot_dir)) dir.create(plot_dir)

cat("Output directories initialized under './outputs/'.\n\n")

# --- 3. Run Pipeline Stages ---

# Stage 1: Load and clean data
cat("----------------------------------------------------------------------\n")
cat("Stage 1: Executing Data Loader & Cleaning...\n")
if (file.exists("01_data_loader.R")) {
  source("01_data_loader.R")
} else {
  stop("CRITICAL ERROR: Missing '01_data_loader.R' script.")
}

# Stage 2: Knowledge Scoring & Z-Scores
cat("----------------------------------------------------------------------\n")
cat("Stage 2: Executing Scoring System...\n")
if (file.exists("02_scoring.R")) {
  source("02_scoring.R")
} else {
  stop("CRITICAL ERROR: Missing '02_scoring.R' script.")
}

# Stage 3: Descriptive Statistics Summary
cat("----------------------------------------------------------------------\n")
cat("Stage 3: Running Descriptive Statistics summaries...\n")
if (file.exists("03_descriptive_stats.R")) {
  source("03_descriptive_stats.R")
} else {
  stop("CRITICAL ERROR: Missing '03_descriptive_stats.R' script.")
}

# Stage 4: Visualizations
cat("----------------------------------------------------------------------\n")
cat("Stage 4: Generating plots...\n")
if (file.exists("04_visualizations.R")) {
  source("04_visualizations.R")
} else {
  stop("CRITICAL ERROR: Missing '04_visualizations.R' script.")
}

# Stage 5: Inferential Statistical Analysis
cat("----------------------------------------------------------------------\n")
cat("Stage 5: Executing Inferential Statistical Analysis...\n")
if (file.exists("05_inferential_stats.R")) {
  source("05_inferential_stats.R")
} else {
  stop("CRITICAL ERROR: Missing '05_inferential_stats.R' script.")
}

cat("======================================================================\n")
cat("Pipeline Executed Successfully!\n")

cat("CSV tables are in: ", csv_dir, "\n")
cat("Plots are in:      ", plot_dir, "\n")
cat("======================================================================\n")
